API Interface 구성하기

Dae-Hwa opened the pull request:
https://github.com/Im-D/your-voice-back/issues/6 구현 사항입니다.

작업 내역 아래의 논의 사항 보시고 의견 부탁드릴게요

**작업 내역**

- `Question`, `QuestionLike` 엔티티 및 DTO 객체 추가했습니다.
- create api 작성했습니다. 서비스와 db 로직 제외한 컨트롤러 작업만 이루어지고 있습니다.
- api 규격은 다음과 같습니다(피드백 부탁드립니다).
    - request
        ```json
        {
            "contents": "test",
            "emoji": "test",
            "createDateTime": "2021-03-03 22:34:00"
        }
        ```
    - response
        ```json
        {
            "data": {
                "id": null,
                "indexNumber": null,
                "contents": "test",
                "emoji": "test",
                "createDateTime": "2021-03-03 22:34:00",
                "questionLikeDTOs": null
            },
            "isSuccess": "success",
            "message": "CREATE SUCCESSFUL"
        }
        ```
        > id와 indexNumber는 추후 업데이트 될 예정입니다.
- 현재 실패하는 경우는 다음과 같습니다.
    - json에 알 수 없는 프로퍼티 가 포함되었을 경우
    - contents가 null or empty일 경우
    - contents가 140글자 제한을 넘어갈 경우
---
**논의 필요 사항**
- 날짜 규격 어떻게 할 건지 정해야 할 것 같습니다.
- data 현재 모양으로 사용할 것인지, 내부에 객체로 한 번 더 감쌀 것인지
- 실패 시 data에 어떤 내용 넣을 것인지
   > 제 입장에선 비어있는 채로 리턴하는게 제일 편하긴 합니다.. 😅

Dae-Hwa committed:
feat: 질문 및 좋아요 entity 객체 추가

Dae-Hwa committed:
fix: Question, QuestionLike DTO 생성

Dae-Hwa committed:
feat: Question create rest api 생성

현재 서비스 로직과 레포지토리는 없는 상태. api request와 response만 존재

Dae-Hwa committed:
fix: Question 도메인 클래스들에 있는 long타입 래퍼 클래스로 교체

null 구분이 가능해야 함

Dae-Hwa committed:
test: QuestionController 테스트에 json 문자열 request 테스트 추가

Dae-Hwa committed:
feat: UnknownRequestPropertiesExceptionHandler 추가

request body json에 정의되지 않은 프로퍼티가 들어있을 경우 에러 발생. 해당 기능을 위한 jackson 설정
common.properties에 추가

SeonHyungJo commented:
@Dae-Hwa 바로 master로 머지인가요?

Dae-Hwa commented:
> @Dae-Hwa 바로 master로 머지인가요?

@SeonHyungJo 네 그럴 생각이었는데, dev로 합쳐야 하나요? 머지되면 리베이스 받아서 dev에서 계속 쓰려고 했습니다.

Dae-Hwa commented:
- [x] Date : ISO 포멧(날짜-시간 사이 공백 대신 T)(미완료)
- [x] data 포멧 그대로(플랫하게)
- [x] fail시 리퀘스트 데이터 그대로 리턴(미완료)
- [x] request별로 DTO 분리(미완료)

BKJang commented on src/main/java/com/imd/yourvoice/common/ValidateExceptionHandler.java:
```diff
@@ -0,0 +1,29 @@
+        ResponseDTO<?> responseDTO = ResponseDTO.builder()
+                .data(ex.getTarget())
+                .message("Validation Error")
+                .isSuccess("fail")
```
message나 isSuccess에 대한 response 데이터는 상수로 정의해두는건 어떨까요?

Dae-Hwa replied:
단순히 상수로 정의될 종류는 아닌 것 같아 어느 수준까지 정의되면 enum으로 적용시킬 예정입니다.

BKJang reviewed (approved):
comment하나 남겨두었습니다. 확인부탁드려요.
미리 승인하겠습니다 👍

Dae-Hwa committed:
feat: MappingJackson2HttpMessageConverterWithCachingRequest추가

Request를 Handler에서 사용하면 body를 읽을 수 없음(InputStream 사용함). 때문에 캐싱해둔 뒤 읽을 수
있도록 잭슨(json파서) 일부 커스텀

Dae-Hwa committed:
fix: ResponseEntityExceptionHandler 중복 오류 수정 및 커스텀한 잭슨컨버터 적용

Dae-Hwa committed:
refactor: autowired 생성자 주입으로 변경

Dae-Hwa committed:
refactor: 클래스 내부 필드 가독성 개선

Dae-Hwa committed:
test: 통합테스트 샘플 삭제

RestTemplate 사용 시 디스패쳐 서블릿을 거치지 않아 리퀘스트 확인 불가. 필요할 경우 MockMvc로 구성하여
통합테스트 필요

Dae-Hwa committed:
fix: 지워지지 않은 테스트 어트리뷰트 추가 삭제

Dae-Hwa committed:
feat: Create 요청을 위한 DTO 분리

Dae-Hwa committed:
fix: 시간 형식 ISO포멧으로 변경. Const클래스 생성하여 중복 제거

Dae-Hwa commented:
@SeonHyungJo @BKJang @dididy 

수정사항 반영 완료 했습니다.

- Date : ISO 포멧(날짜-시간 사이 공백 대신 T)

- request별로 DTO 분리

- fail시 리퀘스트 데이터 그대로 리턴

  까다로웠던 부분인데, JSON파서를 커스텀 해서 해결했습니다.
  간단히 설명하면, 
    - 스프링에 json 리퀘스트가 들어오면 jackson라이브러리로 파싱해주는데 이 때, 리퀘스트 body가 InputStream 형태로 들어있습니다. 
    - InputStream은 I/O에 사용되는데, 스프링에서 사용되는 ServletRequest는 body가 읽기 완료되면 더 이상 호출이 불가능하도록 막아놨습니다. 따라서 재사용이 불가능하여 body만 리퀘스트에 별도로 저장하도록 했습니다.
    - 추후에 별도의 인터셉터나 AOP로 분리하여 구현이 가능한지 확인하고 개선하도록 하겠습니다.

voidmatcha reviewed (approved):
수고 많으셨습니다~

Dae-Hwa committed:
refactor: 커스텀 잭슨 컨버터 필터로 대체

ContentCachingRequestWrapper로 내부의 스트림 캐싱하도록 변경

Dae-Hwa commented:
커스텀한 부분 필터로 대체했습니다.
스프링에서 제공하는 캐싱 객체 사용했습니다.
